File: lab1-22.02.21/src/main/java/testpassword/lab1/services/AdvertService.java
package testpassword.lab1.services;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import testpassword.lab1.models.Advert;
import testpassword.lab1.repos.AdvertRepo;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

@Service public class AdvertService {

    private final AdvertRepo repo;

    @Autowired public AdvertService(AdvertRepo repo) {
        this.repo = repo;
    }

    public boolean add(Advert a) {
        if (a.getUser() == null || !ModeratorService.moderate(a)) return false;
        else {
            this.save(a);
            return true;
        }
    }

    public List<Advert> getAll() {
        return StreamSupport.stream(repo.findAll().spliterator(), false).collect(Collectors.toList());
    }

    public Advert get(long id) { return repo.getByAdvertId(id); }

    @Transactional public void save(Advert a) { this.repo.save(a); }

    @Transactional public void delete(Advert a) { repo.delete(a); }
}

File: lab3-18.05.21/src/main/kotlin/lab3/services/AdvertService.kt
package lab3.services

import lab3.dtos.http.AdvertReq
import lab3.models.Advert
import lab3.models.RoleException
import lab3.repos.AdvertRepo
import lab3.utils.AutoModerator
import lab3.utils.Postman
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service class AdvertService {

    @Autowired private lateinit var repo: AdvertRepo
    @Autowired private lateinit var postman: Postman

    @Transactional infix fun add(advert: Advert) =
        with(AutoModerator(advert)) {
            if (isEmpty()) {
                repo.save(advert)
                true
            } else {
                postman(advert.user.email,
                    "Automatic moderation failed",
                    """
                    Failed automatic moderation cause you use prohibited words:
                    $this
                    Replace them with something else
                    """.trimIndent()
                )
                false
            }
        }

    @Transactional fun changeStatus(advertId: Long, status: Advert.STATUS, comment: String = "") =
        with(get(advertId)) {
            val templateMail: (String) -> Unit = { postman(this.user.email, "Moderation result", it) }
            this.status = status
            when (status) {
                Advert.STATUS.DECLINED -> {
                    repo.delete(this)
                    templateMail("The moderator decided to remove your ad for the following reason:\n$comment")
                }
                Advert.STATUS.ON_MODERATION -> {
                    repo.save(this)
                    templateMail("The moderator decided that you need to modify the ad due to:\n$comment")
                }
                Advert.STATUS.APPROVED -> {
                    publishDate = Date()
                    repo.save(this)
                    templateMail("Congratulations! Ad approved.")
                }
                Advert.STATUS.CLOSED -> {
                    repo.save(this)
                    templateMail("You closed advert")
                }
            }
        }

    @Transactional fun modify(advertId: Long, modified: AdvertReq) {
        val orig = get(advertId)
        modified.fillAdvert(orig)
        modified.status?.let { if (it == Advert.STATUS.CLOSED) orig.status = it else throw RoleException() }
        repo.save(orig)
    }

    fun getAll(status: Advert.STATUS = Advert.STATUS.APPROVED) = repo.findAll().filter { it.status == status }.toList()

    infix operator fun get(id: Long) = repo.getByAdvertId(id)

    infix fun delete(advert: Advert) = repo.delete(advert)
}